File: ios/HoloNet/Module/Forum/Data/ForumThreadRepository.swift
//
//  ForumThreadRepository.swift
//  SWTOR HoloNet
//
//  Created by Ivan Fabijanovic on 17/10/14.
//  Copyright (c) 2014 Ivan Fabijanović. All rights reserved.
//

import UIKit
import Alamofire
import RxSwift
import RxAlamofire
import HTMLReader

protocol ForumThreadRepository {
    func threads(language: ForumLanguage, category: ForumCategory, page: Int) -> Observable<[ForumThread]>
}

class DefaultForumThreadRepository: ForumRepositoryBase, ForumThreadRepository {
    func threads(language: ForumLanguage, category: ForumCategory, page: Int) -> Observable<[ForumThread]> {
        return self.threads(language: language, id: category.id, page: page)
    }
}

extension DefaultForumThreadRepository {
    fileprivate func threads(language: ForumLanguage, id: Int, page: Int) -> Observable<[ForumThread]> {
        return self.manager.rx
            .string(.get, self.url(language: language, id: id, page: page))
            .map {
                let items = self.parse(html: $0)
                if items.isEmpty && self.isMaintenanceResponse($0) {
                    throw ForumError.maintenance
                }
                return items
            }
    }
    
    private func url(language: ForumLanguage, id: Int, page: Int) -> URL {
        let localizedSettings = self.localizedSettings(language: language)
        let string = "\(localizedSettings.forumDisplayUrl)?\(self.settings.categoryQueryParam)=\(id)&\(self.settings.pageQueryParam)=\(page)"
        let url = URL(string: string)
        assert(url != nil)
        return url!
    }
    
    private func parse(html: String) -> [ForumThread] {
        var items = [ForumThread]()
        
        let document = HTMLDocument(string: html)
        let threadNodes = document.nodes(matchingSelector: "table#threadslist tr")
        
        for (index, node) in threadNodes.enumerated() {
            if let thread = self.parseThread(element: node, index: index) {
                items.append(thread)
            }
        }
        
        return items
    }
    
    private func parseThread(element: HTMLElement, index: Int) -> ForumThread? {
        // Id & Title
        let titleElement = element.firstNode(matchingSelector: ".threadTitle")
        let idString = self.parser.linkParameter(linkElement: titleElement, name: self.settings.threadQueryParam)
        let id = idString != nil ? Int(idString!) : nil
        let title = titleElement?.textContent

        // Last post date
        let lastPostDate = element.firstNode(matchingSelector: ".lastpostdate")?.textContent
        
        // Author
        let author = element.firstNode(matchingSelector: ".author")?.textContent
        
        // Replies
        let replies = self.parser.integerContent(element: element.firstNode(matchingSelector: ".resultReplies"))
        
        // Views
        let views = self.parser.integerContent(element: element.firstNode(matchingSelector: ".resultViews"))
        
        // Has Bioware reply & sticky
        var hasBiowareReply = false
        var isSticky = false
        let imageElements = element.nodes(matchingSelector: ".threadLeft img.inlineimg")
        for image in imageElements {
            let src = image["src"]
            if src == nil {
                continue
            }
            
            if src == self.settings.devTrackerIconUrl {
                hasBiowareReply = true
            } else if src == self.settings.stickyIconUrl {
                isSticky = true
            }
        }
        
        if id == nil { return nil }
        if title == nil { return nil }
        if lastPostDate == nil { return nil }
        if author == nil { return nil }
        if replies == nil { return nil }
        if views == nil { return nil }
        
        let finalTitle = title!.stripNewLinesAndTabs().trimSpaces().collapseMultipleSpaces()
        let finalLastPostDate = lastPostDate!.stripNewLinesAndTabs().trimSpaces().collapseMultipleSpaces()
        let finalAuthor = author!.stripNewLinesAndTabs().trimSpaces().collapseMultipleSpaces()
        
        var thread = ForumThread(id: id!, title: finalTitle, lastPostDate: finalLastPostDate, author: finalAuthor, replies: replies!, views: views!, hasBiowareReply: hasBiowareReply, isSticky: isSticky)
        thread.loadIndex = index
        
        return thread
    }
}

File: android/app/src/main/java/com/ifabijanovic/holonet/forum/data/ForumThreadRepository.kt
package com.ifabijanovic.holonet.forum.data

import com.ifabijanovic.holonet.app.model.Settings
import com.ifabijanovic.holonet.forum.language.ForumLanguage
import com.ifabijanovic.holonet.forum.model.ForumCategory
import com.ifabijanovic.holonet.forum.model.ForumMaintenanceException
import com.ifabijanovic.holonet.forum.model.ForumThread
import com.ifabijanovic.holonet.helper.StringHelper
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import org.jsoup.Jsoup
import org.jsoup.nodes.Element

/**
 * Created by Ivan Fabijanovic on 23/03/2017.
 */
interface ForumThreadRepository {
    fun threads(language: ForumLanguage, category: ForumCategory, page: Int): Observable<List<ForumThread>>
}

class DefaultForumThreadRepository(parser: ForumParser, service: ForumService, settings: Settings) : ForumRepositoryBase(parser, service, settings), ForumThreadRepository {
    override fun threads(language: ForumLanguage, category: ForumCategory, page: Int): Observable<List<ForumThread>> {
        val localizedSettings = this.localizedSettings(language)

        return this.service
                .category(localizedSettings.pathPrefix, category.id)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.computation())
                .map { html ->
                    val items = this.parse(html)
                    if (items.isEmpty() && this.isMaintenance(html)) {
                        throw ForumMaintenanceException()
                    }
                    return@map items
                }
    }

    private fun parse(html: String): List<ForumThread>{
        val items = mutableListOf<ForumThread>()

        Jsoup.parse(html)
                .select("table#threadslist tr")
                .mapIndexedNotNullTo(items) { index, node -> this.parseThread(node, index) }

        return items.toList()
    }

    private fun parseThread(element: Element, index: Int): ForumThread? {
        // Id & Title
        val titleElement = element.select(".threadTitle")?.first()
        val idString = this.parser.linkParameter(titleElement, this.settings.threadQueryParam)
        var id: Int?
        try {
            id = idString?.toInt()
        } catch (e: Exception) {
            id = null
        }
        val title = titleElement?.text()

        // Last post date
        val lastPostDate = element.select(".lastpostdate")?.first()?.text()

        // Author
        val author = element.select(".author")?.first()?.text()

        // Replies
        val replies = this.parser.integerContent(element.select(".resultReplies")?.first())

        // Views
        val views = this.parser.integerContent(element.select(".resultViews")?.first())

        // Has Bioware reply & sticky
        var hasBiowareReply = false
        var isSticky = false
        val imageElements = element.select(".threadLeft img.inlineimg")
        imageElements
                .mapNotNull { it.attr("src") }
                .forEach {
                    if (it == this.settings.devTrackerIconUrl) {
                        hasBiowareReply = true
                    } else if (it == this.settings.stickyIconUrl) {
                        isSticky = true
                    }
                }

        if (id == null) { return null }
        if (title == null) { return null }
        if (lastPostDate == null) { return null }
        if (author == null) { return null }
        if (replies == null) { return null }
        if (views == null) { return null }

        val finalTitle = StringHelper(title).stripNewLinesAndTabs().trimSpaces().collapseMultipleSpaces().value
        val finalLastPostDate = StringHelper(lastPostDate).stripNewLinesAndTabs().trimSpaces().collapseMultipleSpaces().value
        val finalAuthor = StringHelper(author).stripNewLinesAndTabs().trimSpaces().collapseMultipleSpaces().value

        return ForumThread(id, finalTitle, finalLastPostDate, finalAuthor, replies, views, hasBiowareReply, isSticky, false, index)
    }
}
